controller_manager/src/controller_manager.cpp

```diff
@@ -524,42 +604,388 @@ controller_interface::return_type ControllerManager::switch_controller(
-                                 const ControllerSpec controller,
-                                 std::vector<std::string> & request_interface_list) {
-    auto command_interface_config = controller.c->command_interface_configuration();
-    std::vector<std::string> command_interface_names = {};
-    if (command_interface_config.type == controller_interface::interface_configuration_type::ALL)
+  // lock controllers
+  std::lock_guard<std::recursive_mutex> guard(rt_controllers_wrapper_.controllers_lock_);
+
+  const std::vector<ControllerSpec> & controllers = rt_controllers_wrapper_.get_updated_list(guard);
+
+  // if a preceding controller is deactivated, all first-level controllers should be switched 'from'
+  // chained mode
+  for (const auto & controller : controllers)
+  {
+    // get pointers to places in stop and start lists (start/stop lists have changed)
+    auto stop_list_it = std::find(stop_request_.begin(), stop_request_.end(), controller.info.name);
+
+    if (stop_list_it != stop_request_.end())
     {
-      command_interface_names = resource_manager_->available_command_interfaces();
+      // if controller is not active then skip adding following-controllers to "from" chained mode
+      // request
+      if (!is_controller_active(controller.c))
+      {
+        RCLCPP_DEBUG(
+          get_logger(),
+          "Controller with name '%s' can not be deactivated since is not active. "
+          "The controller will be removed from the list later."
+          "Skipping adding following controllers to 'from' chained mode request.",
+          controller.info.name.c_str());
+        break;
+      }
+
+      for (const auto & cmd_itf_name : controller.c->command_interface_configuration().names)
+      {
+        // controller that 'cmd_tf_name' belongs to
+        ControllersListIterator following_ctrl_it;
+        if (command_interface_is_reference_interface_of_controller(
+              cmd_itf_name, controllers, following_ctrl_it))
+        {
+          // currently iterated "controller" is preceding controller --> add following controller
+          // with matching interface name to "from" chained mode list (if not already in it)
+          if (
+            std::find(
+              from_chained_mode_request_.begin(), from_chained_mode_request_.end(),
+              following_ctrl_it->info.name) == from_chained_mode_request_.end())
+          {
+            from_chained_mode_request_.push_back(following_ctrl_it->info.name);
+            RCLCPP_DEBUG(
+              get_logger(), "Adding controller '%s' in 'from chained mode' request.",
+              following_ctrl_it->info.name.c_str());
+          }
+        }
+      }
     }
-    if (
-      command_interface_config.type ==
-      controller_interface::interface_configuration_type::INDIVIDUAL)
+  }
+
+  // check if controllers should be switched 'to' chained mode when controllers are activated
+  const auto check_following_controllers_for_activate =
```

Comment:
this is a 120 lines function that is also recursive, posted in here as a lambda...

please put this in a separate file at least with proper docs and later we can add some testing to it

yeah it is ok for it to have 10 arguments

Reply:
Done.

Didn't put into separate file but as the member of controller manager, so we don't have to have so many arguments. Would this also work?

Reply:
ok for now, that is one step closer to something testable, we don't break API by removing stuff from the private section